include/cuco/detail/static_multimap.inl

```diff
@@ -0,0 +1,1153 @@
+
+      thread_num_matches += equals;
+
+      if (g.any(slot_is_empty)) {
+        if ((not found_match) && (g.thread_rank() == 0)) { thread_num_matches++; }
+        break;
+      }
+
+      current_slot = next_slot(current_slot);
+    }
+  } else {
+    while (true) {
+      auto slot_contents = *reinterpret_cast<cuco::pair_type<Key, Value> const*>(current_slot);
+
+      auto const slot_is_empty = (slot_contents.first == this->get_empty_key_sentinel());
+
+      auto const equals = not slot_is_empty and pair_equal(slot_contents, pair);
+
+      thread_num_matches += equals;
+
+      if (g.any(slot_is_empty)) { break; }
+
+      current_slot = next_slot(current_slot);
+    }
+  }
+}
+
+template <typename Key,
+          typename Value,
+          class ProbeSequence,
+          cuda::thread_scope Scope,
+          typename Allocator>
+template <uint32_t buffer_size,
+          bool is_vector_load,
+          bool is_outer,
+          typename CG,
+          typename atomicT,
+          typename OutputIt,
+          typename KeyEqual>
+__device__ std::enable_if_t<is_vector_load, void>
+static_multimap<Key, Value, ProbeSequence, Scope, Allocator>::device_view::retrieve(
+  const unsigned int activemask,
+  CG const& g,
+  Key const& k,
+  uint32_t* warp_counter,
+  value_type* output_buffer,
+  atomicT* num_matches,
+  OutputIt output_begin,
+  KeyEqual key_equal) noexcept
+{
+  const uint32_t warp_lane_id = threadIdx.x % 32;
+  const uint32_t cg_lane_id   = g.thread_rank();
+
+  auto current_slot = initial_slot(g, k);
+
+  bool running = true;
+  if constexpr (is_outer) {
+    bool found_match = false;
+
+    while (__any_sync(activemask, running)) {
+      if (running) {
+        pair<Key, Value> arr[2];
+        if constexpr (sizeof(Key) == 4) {
+          auto const tmp = *reinterpret_cast<uint4 const*>(current_slot);
+          memcpy(&arr[0], &tmp, 2 * sizeof(pair<Key, Value>));
+        } else {
+          auto const tmp = *reinterpret_cast<ulonglong4 const*>(current_slot);
+          memcpy(&arr[0], &tmp, 2 * sizeof(pair<Key, Value>));
+        }
+
+        auto const first_slot_is_empty  = (arr[0].first == this->get_empty_key_sentinel());
+        auto const second_slot_is_empty = (arr[1].first == this->get_empty_key_sentinel());
+        auto const first_equals         = (not first_slot_is_empty and key_equal(arr[0].first, k));
+        auto const second_equals        = (not second_slot_is_empty and key_equal(arr[1].first, k));
+        auto const first_exists         = g.ballot(first_equals);
+        auto const second_exists        = g.ballot(second_equals);
+
+        if (first_exists or second_exists) {
+          found_match = true;
+
+          auto num_first_matches  = __popc(first_exists);
+          auto num_second_matches = __popc(second_exists);
+
+          uint32_t output_idx;
+          if (0 == cg_lane_id) {
+            output_idx = atomicAdd(warp_counter, (num_first_matches + num_second_matches));
```

Comment:
Should we use `cuda::atomic` and `fetch_add` here instead?

Reply:
That's what I did in my initial implementation, the problem is that declaring static shared memory `cuda::atomic` variables will introduce tons of warnings during compilation:
```
warning: dynamic initialization is not supported for a function-scope static __shared__ variable within a __device__/__global__ function
```
This should be a bug and we can use `cuda::atomic` once this issue is fixed. I've added `TODO` comments as a reminder.